File: mobile/ios/Sources/ListingsViewModel.swift
import Foundation
import SwiftUI

@MainActor
final class ListingsViewModel: ObservableObject {
    // Results
    @Published var listings: [Listing] = []
    @Published var isLoading = false
    @Published var errorMessage: String?

    // Search inputs (bound to the search header)
    @Published var locationQuery = ""
    @Published var useDates = false
    @Published var checkIn = Calendar.current.startOfDay(for: Date())
    @Published var checkOut = Calendar.current.date(byAdding: .day, value: 3, to: Date()) ?? Date()
    @Published var guests = 1

    // Region + sort (bound to the chips row + sort menu under the search field)
    /// Curated browse regions with counts, from `GET /api/local/regions`.
    @Published var regions: [RegionFacet] = []
    /// Active region filter; `nil` means "All" (no region constraint).
    @Published var selectedRegion: String?
    /// Active sort order; defaults to the backend's "recommended".
    @Published var sort: ListingSort = .recommended

    // Discovery filters (set in the Filters sheet) — applied on top of the
    // region/search/date filters above.
    /// Amenities the listing must ALL have. Sent comma-joined as `amenities`.
    @Published var selectedAmenities: Set<String> = []
    /// Single property-type filter (e.g. "Villa"); `nil` means "Any type".
    @Published var selectedPropertyType: String?

    /// Property types offered in the Filters sheet, matching the backend's
    /// case-insensitive `propertyType` values.
    let propertyTypes: [String] = ["Apartment", "Chalet", "House", "Villa"]

    /// True once the user has run a filtered search — used to switch the empty
    /// state copy from "seed the database" to "no stays match your search".
    @Published var isFiltered = false

    // Place autocomplete (web + Android parity) — suggestions for the location
    // field from `GET /api/local/places`, refreshed (debounced) as the user types.
    @Published var placeSuggestions: [String] = []
    /// Monotonic token so a stale suggestions task can detect it was superseded.
    private var placeSuggestToken = 0

    /// Debounced place-suggestions refresh for the location search field. Called
    /// from the header whenever `locationQuery` changes (and when the field gains
    /// focus with an empty query, which returns the curated popular destinations).
    func suggestPlaces(debounced: Bool = true) {
        placeSuggestToken += 1
        let token = placeSuggestToken
        let query = locationQuery
        Task {
            if debounced { try? await Task.sleep(nanoseconds: 300_000_000) }
            guard token == placeSuggestToken else { return }
            let places = await SupabaseService.shared.fetchPlaceSuggestions(query: query)
            guard token == placeSuggestToken else { return }
            placeSuggestions = places
        }
    }

    /// Hides the suggestion list (on submit / selection / clear).
    func clearPlaceSuggestions() {
        placeSuggestToken += 1
        placeSuggestions = []
    }

    private static let dayFormatter: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.dateFormat = "yyyy-MM-dd"
        return f
    }()

    /// Short human label for the currently-selected range, e.g. "Jun 12 → Jun 15".
    /// `nil` when no dates are active so the UI can show its placeholder.
    var dateRangeLabel: String? {
        guard useDates else { return nil }
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US")
        f.dateFormat = "MMM d"
        return "\(f.string(from: checkIn)) → \(f.string(from: checkOut))"
    }

    /// Apply a range chosen in the `DateRangePicker`. A non-nil pair turns dates
    /// on; clearing (nil) turns them off and resets to the defaults.
    func applyDateRange(checkIn newIn: Date?, checkOut newOut: Date?) {
        if let newIn, let newOut {
            checkIn = newIn
            checkOut = newOut
            useDates = true
        } else {
            useDates = false
            checkIn = Calendar.current.startOfDay(for: Date())
            checkOut = Calendar.current.date(byAdding: .day, value: 3, to: Date()) ?? Date()
        }
    }

    /// Initial / pull-to-refresh load — respects the current filter state so a
    /// refresh after searching keeps the same results.
    func load() async {
        await fetch(filtered: isFiltered)
    }

    /// Run a search with the current header inputs.
    func search() async {
        await fetch(filtered: true)
    }

    /// Reset the header to its defaults and reload the full catalogue. Also
    /// clears the region chip + sort so "Clear" returns to a truly blank slate.
    func clear() async {
        locationQuery = ""
        useDates = false
        guests = 1
        checkIn = Calendar.current.startOfDay(for: Date())
        checkOut = Calendar.current.date(byAdding: .day, value: 3, to: Date()) ?? Date()
        selectedRegion = nil
        sort = .recommended
        selectedAmenities = []
        selectedPropertyType = nil
        await fetch(filtered: false)
    }

    /// Apply the discovery filters chosen in the Filters sheet and refetch,
    /// keeping the active region / search / dates. Treated as a filtered fetch so
    /// the empty state reads "no stays match" rather than the seed prompt.
    func applyFilters() async {
        await fetch(filtered: true)
    }

    /// Reset just the discovery filters (amenities + property type) and refetch,
    /// leaving the region / search / dates untouched.
    func clearFilters() async {
        selectedAmenities = []
        selectedPropertyType = nil
        await load()
    }

    /// Whether any discovery filter (amenities or property type) is set. Drives
    /// the Filters button's "active" badge.
    var hasDiscoveryFilters: Bool {
        !selectedAmenities.isEmpty || selectedPropertyType != nil
    }

    /// Number of active discovery filters, shown as a count badge on the Filters
    /// button (each selected amenity + a chosen property type each count once).
    var discoveryFilterCount: Int {
        selectedAmenities.count + (selectedPropertyType == nil ? 0 : 1)
    }

    /// "Search this area" on the map: refetch listings inside `box`, combined with
    /// any active region / search / date / discovery filters. The box is one-shot
    /// (not stored), so later region/sort/search changes return to the full set.
    func searchArea(_ box: BBox) async {
        await fetch(filtered: true, bbox: box)
    }

    /// Select a region chip (or `nil` for "All") and refetch. Region filtering is
    /// independent of the search header, so it keeps any active location/dates.
    func selectRegion(_ region: String?) async {
        guard selectedRegion != region else { return }
        selectedRegion = region
        await load()
    }

    /// Change the sort order and refetch, keeping all other filters.
    func applySort(_ newSort: ListingSort) async {
        guard sort != newSort else { return }
        sort = newSort
        await load()
    }

    /// Load the curated browse regions (with counts) for the chips row. Best
    /// effort — a failure just leaves the chips at "All" only.
    func loadRegions() async {
        if let fetched = try? await SupabaseService.shared.fetchRegions() {
            regions = fetched
        }
    }

    /// Whether any filter beyond the default is active (region, sort, the search
    /// header, or a discovery filter). Drives the empty-state copy + the "Clear
    /// search" button.
    var anyFilterActive: Bool {
        isFiltered || selectedRegion != nil || sort != .recommended || hasDiscoveryFilters
    }

    private func fetch(filtered: Bool, bbox: BBox? = nil) async {
        isLoading = true
        errorMessage = nil
        isFiltered = filtered
        do {
            let dates = filtered && useDates
            listings = try await SupabaseService.shared.fetchListings(
                location: filtered ? locationQuery : nil,
                region: selectedRegion,
                guests: filtered && guests > 0 ? guests : nil,
                checkIn: dates ? Self.dayFormatter.string(from: checkIn) : nil,
                checkOut: dates ? Self.dayFormatter.string(from: checkOut) : nil,
                sort: sort,
                propertyType: selectedPropertyType,
                amenities: Array(selectedAmenities),
                bbox: bbox
            )
            if listings.isEmpty {
                errorMessage = anyFilterActive
                    ? "No stays match your filters. Try a different area or clearing them."
                    : "No listings found yet. Seed the database to see stays here."
            }
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }
}

File: mobile/android/app/src/main/java/com/quickin/app/ListingsViewModel.kt
package com.quickin.app

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class ListingsUiState(
    val isLoading: Boolean = false,
    val listings: List<Listing> = emptyList(),
    val error: String? = null,
    /** The filters used for the current results (so the header can reflect / clear them). */
    val query: ListingQuery = ListingQuery(),
    /** Curated browse regions with counts (`GET /api/local/regions`) for the chip row. */
    val regions: List<Region> = emptyList()
)

/**
 * The "More from this host" rail on the listing detail (`GET /api/local/listings?host=<id>`).
 * [hostId] is the host the rail was loaded for, so a stale result for a previously-opened
 * listing can be ignored. The current listing is excluded by the screen, not here.
 */
data class MoreFromHostUiState(
    val hostId: String? = null,
    val listings: List<Listing> = emptyList()
)

/**
 * State for the natural-language ("Ask AI") search on the explore screen (Section 10,
 * `POST /api/local/ai/search`). This is an ADDITIONAL search mode that sits alongside the regular
 * filter search — it never mutates [ListingsUiState]. When [active] is true the screen shows the
 * AI [results] plus the parsed [filters] as chips; clearing it returns to the regular feed.
 *
 * [query] is the prose the guest typed (echoed back on the bar), [isSearching] drives the spinner,
 * and [error] surfaces a friendly note when the AI is unavailable.
 */
data class AiSearchUiState(
    val active: Boolean = false,
    val isSearching: Boolean = false,
    val query: String = "",
    val filters: AiSearchFilters = AiSearchFilters(),
    val results: List<Listing> = emptyList(),
    val error: String? = null
)

class ListingsViewModel : ViewModel() {
    private val _state = MutableStateFlow(ListingsUiState())
    val state: StateFlow<ListingsUiState> = _state.asStateFlow()

    // Natural-language ("Ask AI") search — an additional mode over the regular filter search.
    private val _aiSearch = MutableStateFlow(AiSearchUiState())
    val aiSearch: StateFlow<AiSearchUiState> = _aiSearch.asStateFlow()

    // "More from this host" rail for the currently-open listing detail.
    private val _hostListings = MutableStateFlow(MoreFromHostUiState())
    val hostListings: StateFlow<MoreFromHostUiState> = _hostListings.asStateFlow()

    // A listing fetched for an incoming deep link (https://…/explore/{id} or quickin://explore/{id}).
    // MainApp observes this and opens the detail; null once consumed (or when the fetch failed).
    private val _deepLinkListing = MutableStateFlow<Listing?>(null)
    val deepLinkListing: StateFlow<Listing?> = _deepLinkListing.asStateFlow()

    // Place-autocomplete suggestions for the Explore search location field
    // (`GET /api/local/places?q=…`). Debounced; cleared when a suggestion is chosen or the query is short.
    private val _placeSuggestions = MutableStateFlow<List<String>>(emptyList())
    val placeSuggestions: StateFlow<List<String>> = _placeSuggestions.asStateFlow()
    private var placeSuggestJob: Job? = null

    init {
        load()
        loadRegions()
    }

    /**
     * Resolves a deep-linked listing by id (`GET /api/local/listings/:id`) and publishes it on
     * [deepLinkListing] for the UI to open. A missing/invalid id or a fetch failure is silently
     * ignored so a garbage link just leaves the app where it was.
     */
    fun openListingById(id: String) {
        if (id.isBlank()) return
        viewModelScope.launch {
            SupabaseService.fetchListing(id)?.let { _deepLinkListing.value = it }
        }
    }

    /** Consumed by the UI once the deep-linked listing has been opened. */
    fun clearDeepLinkListing() {
        _deepLinkListing.value = null
    }

    /**
     * Loads the host's other listings for the detail's "More from this host" rail. Clears the
     * rail immediately (so a different host's stays never flash) and refetches; a blank
     * [hostId] just clears it. Best-effort — failures yield an empty rail (the section hides).
     */
    fun loadHostListings(hostId: String?) {
        val id = hostId?.takeIf { it.isNotBlank() }
        _hostListings.value = MoreFromHostUiState(hostId = id)
        if (id == null) return
        viewModelScope.launch {
            val listings = SupabaseService.fetchHostListings(id)
            // Ignore a result that arrived after the user opened a different host's listing.
            if (_hostListings.value.hostId == id) {
                _hostListings.value = MoreFromHostUiState(hostId = id, listings = listings)
            }
        }
    }

    /** Clears the "More from this host" rail (on leaving the detail screen). */
    fun clearHostListings() {
        _hostListings.value = MoreFromHostUiState()
    }

    /**
     * Runs a natural-language search (`POST /api/local/ai/search`): the AI parses [query] into
     * structured filters and returns matching listings. This is an additional mode — it never
     * touches the regular filter search/feed. A blank [query] just clears AI mode.
     */
    fun aiSearch(query: String) {
        val q = query.trim()
        if (q.isBlank()) {
            clearAiSearch()
            return
        }
        if (_aiSearch.value.isSearching) return
        _aiSearch.value = AiSearchUiState(active = true, isSearching = true, query = q)
        viewModelScope.launch {
            try {
                val result = SupabaseService.aiSearch(q)
                _aiSearch.value = AiSearchUiState(
                    active = true,
                    isSearching = false,
                    query = q,
                    filters = result.filters,
                    results = result.listings,
                    error = if (result.listings.isEmpty()) "No stays matched. Try rephrasing your search." else null
                )
            } catch (e: Exception) {
                _aiSearch.value = AiSearchUiState(
                    active = true,
                    isSearching = false,
                    query = q,
                    error = e.message ?: "AI search isn't available right now."
                )
            }
        }
    }

    /** Exits AI search mode and returns the screen to the regular filtered feed. */
    fun clearAiSearch() {
        _aiSearch.value = AiSearchUiState()
    }

    /**
     * Fetches place suggestions for the Explore location typeahead (`GET /api/local/places?q=…`),
     * debounced ~250ms so it fires once the user pauses typing. A short (<2 char) query clears the
     * list. Best-effort — a failed lookup just yields no suggestions.
     */
    fun suggestPlaces(query: String) {
        val q = query.trim()
        placeSuggestJob?.cancel()
        if (q.length < 2) {
            _placeSuggestions.value = emptyList()
            return
        }
        placeSuggestJob = viewModelScope.launch {
            kotlinx.coroutines.delay(250)
            val results = try { PlacesService.suggest(q) } catch (e: Exception) { emptyList() }
            _placeSuggestions.value = results
        }
    }

    /** Clears the place-suggestion dropdown (a suggestion was chosen, or the field lost relevance). */
    fun clearPlaceSuggestions() {
        placeSuggestJob?.cancel()
        _placeSuggestions.value = emptyList()
    }

    /** Re-runs the fetch with the current filters (used for Retry). */
    fun load() = fetch(_state.value.query)

    /**
     * Runs a new search with the given filters (preserves the active region, sort, and the
     * discovery filters — property type + amenities). A fresh text/date search drops any
     * pinned map [bbox] so results aren't silently constrained to an old viewport.
     */
    fun search(query: ListingQuery) {
        val current = _state.value.query
        fetch(
            query.copy(
                region = current.region,
                sort = current.sort,
                propertyType = current.propertyType,
                amenities = current.amenities,
                bbox = null
            )
        )
    }

    /**
     * Applies the discovery filters from the Filters sheet (property type + amenities) and
     * refetches, keeping the rest of the search (text / dates / guests / region / sort). Clears
     * any pinned map [bbox] so the new filters search the whole region, not the old viewport.
     */
    fun applyFilters(propertyType: String?, amenities: Set<String>) {
        fetch(
            _state.value.query.copy(
                propertyType = propertyType?.takeIf { it.isNotBlank() },
                amenities = amenities,
                bbox = null
            )
        )
    }

    /** Clears only the discovery filters (property type + amenities), keeping the rest of the search. */
    fun clearFilters() {
        val current = _state.value.query
        if (current.propertyType == null && current.amenities.isEmpty()) return
        fetch(current.copy(propertyType = null, amenities = emptySet()))
    }

    /**
     * Re-queries listings within the map's current visible viewport ("Search this area"),
     * combined with the active filters. [bbox] is "minLng,minLat,maxLng,maxLat"; a blank value
     * clears the viewport constraint.
     */
    fun searchArea(bbox: String) {
        fetch(_state.value.query.copy(bbox = bbox.takeIf { it.isNotBlank() }))
    }

    /**
     * Switches the selected region chip and refetches, keeping the rest of the search
     * (text / dates / guests / sort / discovery filters). A null [region] clears the region
     * filter ("All"). Also drops any pinned map [bbox] so the region change isn't constrained
     * to an old viewport.
     */
    fun selectRegion(region: String?) {
        fetch(_state.value.query.copy(region = region?.takeIf { it.isNotBlank() }, bbox = null))
    }

    /** Switches the sort order and refetches, keeping all other filters. */
    fun setSort(sort: ListingSort) {
        if (_state.value.query.sort == sort) return
        fetch(_state.value.query.copy(sort = sort))
    }

    /** Resets all filters and reloads the full, unfiltered list. */
    fun clear() = fetch(ListingQuery())

    /** Loads the curated browse regions for the chip row (best-effort; failures yield no chips). */
    private fun loadRegions() {
        viewModelScope.launch {
            val regions = SupabaseService.fetchRegions()
            _state.value = _state.value.copy(regions = regions)
        }
    }

    private fun fetch(query: ListingQuery) {
        _state.value = _state.value.copy(isLoading = true, error = null, query = query)
        viewModelScope.launch {
            try {
                val listings = SupabaseService.fetchListings(query)
                _state.value = _state.value.copy(
                    isLoading = false,
                    listings = listings,
                    query = query,
                    error = if (listings.isEmpty()) {
                        if (query.isActive) "No stays match your search. Try different dates or guests."
                        else "No listings found yet. Seed the database to see stays here."
                    } else null
                )
            } catch (e: Exception) {
                _state.value = _state.value.copy(
                    isLoading = false,
                    listings = emptyList(),
                    error = e.message ?: "Something went wrong."
                )
            }
        }
    }
}
